### ISSUE TITLE
Find way to speed up binance historical trade queries

### ISSUE THREAD

**Issue by LefterisJP** [rotki/rotki#1073]
## Problem Definition

Binance historical queries need you to iterate over every single possible market that binance offers:
https://github.com/binance-exchange/binance-official-api-docs/blob/master/rest-api.md#account-trade-list-user_data

This is slow.

There seems to be no endpoint that offers a way to query all pairs in one go.

## Task

Find a way to improve the speed of this query.

**Comment by aspiers** [comment]
Perhaps it could be limited at least to only pairs including tokens for which Rotki knows the user has (or had) non-zero balances?

**Comment by aspiers** [comment]
Or maybe a better option would be simply to submit a request to Binance to fix their API!  Hopefully they'll pay more attention to you than to random API users.

**Comment by LefterisJP** [comment]
They don't :( ... believe me I tried.

> Perhaps it could be limited at least to only pairs including tokens for which Rotki knows the user has (or had) non-zero balances?

Thing is we would need to go through all trade history to see this as they may have no balance now, but may have had before. So then it becomes a chicken and egg problem.

**Comment by aspiers** [comment]
[**@LefterisJP**](https://github.com/LefterisJP) commented on [February 25, 2021 7:38 PM](https://github.com/rotki/rotki/issues/1073#issuecomment-786151573 "2021-02-25T19:38:13Z - Replied by Github Reply Comments"):
> They don't :( ... believe me I tried.

:-(

> > Perhaps it could be limited at least to only pairs including tokens for which Rotki knows the user has (or had) non-zero balances?
> 
> Thing is we would need to go through all trade history to see this as they may have no balance now, but may have had before. So then it becomes a chicken and egg problem.

Not necessarily.  You could just iterate for coins they currently have, and display a warning saying "if you are missing trades for coins where you currently have zero balance, enter the coin here and we'll search for it".  It's definitely not ideal, but maybe it's better than the risk of getting banned or rate-limited by Binance's API due to spam.

Or even if you ignore those ideas, it would at least be worth iterating *first* over pairs where the user currently has non-zero balance, and then scanning the others more slowly in the background.

**Comment by LefterisJP** [comment]
These scans happen always on the background. And with the "show cache" first mechanism we are introducing in the new release this should already feel better.

**Comment by LefterisJP** [comment]
I want to do something for it. The idea about querying specific markets and allowing the user to specify more seems doable and would solve the problem.

Worse UX (not autodetecting everything -- unless user specifies all markets) but much much faster speed and no more rate limits or ip bans.

**Comment by a7e8** [comment]
Why don't you just get the orders of the user first in order to get the markets for the trades? https://binance-docs.github.io/apidocs/spot/en/#all-orders-user_data

Currently you're querying for markets the user has never even traded in so of course you're hitting API limits.

**Comment by LefterisJP** [comment]
> Why don't you just get the orders of the user first in order to get the markets for the trades? https://binance-docs.github.io/apidocs/spot/en/#all-orders-user_data
> 
> Currently you're querying for markets the user has never even traded in so of course you're hitting API limits.

Because their API is terrible. The orders endpoint has the same problem as the trades endpoint. It needs a "symbol" argument. So we would need to do exactly the same thing. We have contacted binance multiple times for this and they do not plan to fix this. It is the only exchange having this ridiculous limitation.

So the only thing we can do is let the user specify markets to query.

**Comment by a7e8** [comment]
> > Why don't you just get the orders of the user first in order to get the markets for the trades? https://binance-docs.github.io/apidocs/spot/en/#all-orders-user_data
> > Currently you're querying for markets the user has never even traded in so of course you're hitting API limits.
> 
> Because their API is terrible. The orders endpoint has the same problem as the trades endpoint. It needs a "symbol" argument. So we would need to do exactly the same thing. We have contacted binance multiple times for this and they do not plan to fix this. It is the only exchange having this ridiculous limitation.
> 
> So the only thing we can do is let the user specify markets to query.

Yes you're right, I was in a rush so missed this detail. They're also a pain to export CSV files too, only allowing 3 months at a time and having to download different types of data from different places.
